##Import Libraries

In [3]:
import pandas as pd
import numpy as np


##Load Dataset

In [1]:
from pathlib import Path
import os
import sys

sys.path.append(str(Path(os.getcwd()).parents[1]))
from Helper_Functions import common_utils
directory_to_extract_to = os.getcwd()

common_utils.load_data_from_one_drive(directory_to_extract_to, "recommendation_paths", "movie_recommendation")

Extracting : 100%|██████████| 2/2 [00:02<00:00,  1.02s/it]


In [4]:
credits = pd.read_csv("tmdb_5000_credits.csv")
movies = pd.read_csv("tmdb_5000_movies.csv")
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [5]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

##Data Exploration

Keep only id, title and overview columns as they are for further processing

In [43]:
credits.isnull().sum()

movie_id    0
title       0
dtype: int64

In [42]:
credits.drop(credits.columns[[2, 3]], axis = 1, inplace = True)

In [48]:
movies.isnull().sum()

id          0
overview    0
dtype: int64

In [45]:
movies.drop(movies.columns[[0,1,2,4,5,6,8,9,10,11,12,13,14,15,16,17,18,19]], axis = 1, inplace = True)

In [47]:
movies = movies.dropna()

##Preprocessing

In [49]:
print("Credits:",credits.shape)
print("Movies Dataframe:",movies.shape)

Credits: (4803, 2)
Movies Dataframe: (4800, 2)


In [50]:
credits_column_renamed = credits.rename(index=str, columns={"movie_id": "id"})
movies_merge = movies.merge(credits_column_renamed, on='id')
print(movies_merge.head())

       id                                           overview  \
0   19995  In the 22nd century, a paraplegic Marine is di...   
1     285  Captain Barbossa, long believed to be dead, ha...   
2  206647  A cryptic message from Bond’s past sends him o...   
3   49026  Following the death of District Attorney Harve...   
4   49529  John Carter is a war-weary, former military ca...   

                                      title  
0                                    Avatar  
1  Pirates of the Caribbean: At World's End  
2                                   Spectre  
3                     The Dark Knight Rises  
4                               John Carter  


In [51]:
movies_cleaned['overview']

0       In the 22nd century, a paraplegic Marine is di...
1       Captain Barbossa, long believed to be dead, ha...
2       A cryptic message from Bond’s past sends him o...
3       Following the death of District Attorney Harve...
4       John Carter is a war-weary, former military ca...
                              ...                        
4794    El Mariachi just wants to play his guitar and ...
4795    A newlywed couple's honeymoon is upended by th...
4796    "Signed, Sealed, Delivered" introduces a dedic...
4797    When ambitious New York attorney Sam is sent t...
4798    Ever since the second grade when he first saw ...
Name: overview, Length: 4799, dtype: object

##Applying vectorization

Convert a collection of raw documents to a matrix of TF-IDF features.

In [52]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfv = TfidfVectorizer(min_df=3,  max_features=None,
            strip_accents='unicode', analyzer='word',token_pattern=r'\w{1,}',
            ngram_range=(1, 3),
            stop_words = 'english')

In [53]:
# Fitting the TF-IDF on the 'overview' text
tfv_matrix = tfv.fit_transform(movies_cleaned['overview'])
print(tfv_matrix)
print(tfv_matrix.shape)

  (0, 148)	0.3091371723189815
  (0, 1670)	0.27815413610097817
  (0, 431)	0.2108413295939603
  (0, 7055)	0.2686774882613109
  (0, 6447)	0.2566772823097974
  (0, 3582)	0.21787716706422044
  (0, 9392)	0.24143973860066492
  (0, 5907)	0.17991690004805883
  (0, 9716)	0.24435186815985976
  (0, 6543)	0.29591523792195684
  (0, 5972)	0.27473495736009795
  (0, 2634)	0.28189942265833556
  (0, 5658)	0.26104797888037473
  (0, 1514)	0.20118105634370556
  (0, 147)	0.3091371723189815
  (1, 1810)	0.36794361320381896
  (1, 7159)	0.3031053671236092
  (1, 2916)	0.30082340684198766
  (1, 9608)	0.3355244901637141
  (1, 2848)	0.21555947963880698
  (1, 2872)	0.3232534752773416
  (1, 4205)	0.3080331741124809
  (1, 5263)	0.13328087934556093
  (1, 1806)	0.21045212959628706
  (1, 2318)	0.21891500721538523
  :	:
  (4798, 671)	0.15980111581570794
  (4798, 2360)	0.15010143939644302
  (4798, 677)	0.13488458793471353
  (4798, 3736)	0.14858681983687555
  (4798, 3480)	0.14339731171521328
  (4798, 1252)	0.1517283521734410

Sigmoid_kernel will output 4799 sigmoid values for each of the 4799 movies we have in our database. Higher the sigmoid values for a particular movie, higher the chances of it being recommended.

In [54]:
from sklearn.metrics.pairwise import sigmoid_kernel

# Compute the sigmoid kernel
sig = sigmoid_kernel(tfv_matrix, tfv_matrix)
print(sig.shape)

(4799, 4799)


##Generating Recommendations

In [55]:
# Reverse mapping of indices and movie titles
indices = pd.Series(movies_cleaned.index, index=movies_cleaned['original_title']).drop_duplicates()

In [62]:
def give_recomendations(title, sig=sig):
    idx = indices[title]
    # Get the pairwsie similarity scores
    sig_scores = list(enumerate(sig[idx]))
    # Sort the movies
    sig_scores = sorted(sig_scores, key=lambda x: x[1], reverse=True)
    # Scores of the 10 most similar movies
    sig_scores = sig_scores[1:11]
    movie_indices = [i[0] for i in sig_scores]
    return movies_cleaned['original_title'].iloc[movie_indices]

In [63]:
print(give_recomendations('Avatar'))

1341                Obitaemyy Ostrov
634                       The Matrix
3603                       Apollo 18
2130                    The American
775                        Supernova
529                 Tears of the Sun
151                          Beowulf
311     The Adventures of Pluto Nash
847                         Semi-Pro
942                 The Book of Life
Name: original_title, dtype: object
